orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/util/ArtifactResolver.groovy

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright 2017 Google, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.orca.pipeline.util
+
+import com.netflix.spinnaker.kork.artifacts.model.Artifact
+import com.netflix.spinnaker.kork.artifacts.model.ExpectedArtifact
+import groovy.util.logging.Slf4j
+
+import java.lang.reflect.Field
+
+@Slf4j
+class ArtifactResolver {
+
+  static void resolveArtifacts(Map pipeline) {
+    List<Artifact> resolvedArtifacts = []
+    if (pipeline.trigger.expectedArtifacts) {
+      // Resolve artifacts based on expectedArtifacts.
+      List<Artifact> receivedArtifacts = pipeline.receivedArtifacts ?: []
+
+      // Split up expected artifacts on whether they're satisfied or not.
+      def (hasMatch, needsLookup) = pipeline.trigger.expectedArtifacts.split { e ->
+        receivedArtifacts.any { a -> expectedMatch((ExpectedArtifact) e, (Artifact) a) }
+      }
+
+      // If satisfied, just find the artifact that matched.
+      hasMatch?.each { e -> resolvedArtifacts << (receivedArtifacts.find { a -> expectedMatch((ExpectedArtifact) e, (Artifact) a) }) }
+
+      // Otherwise, resolve artifact based on the missing fieldType and missingPolicy.
+      needsLookup?.each { e ->
+        receivedArtifacts.each { a ->
+          e.fields?.each { field ->
+            if (!lookupField((ExpectedArtifact.ArtifactField) field, (Artifact) a)) {
+              switch (field.getFieldType()) {
+                case ExpectedArtifact.ArtifactField.FieldType.MUST_MATCH:
+                  throw new ArtifactResolutionException(
+                      "Received artifact value did not match expected artifact for fieldName: ${field.getFieldName()}")
+                  break
+                case ExpectedArtifact.ArtifactField.FieldType.FIND_IF_MISSING:
+                  switch (field.getMissingPolicy()) {
+                    case ExpectedArtifact.ArtifactField.MissingPolicy.FAIL_PIPELINE:
+                      throw new ArtifactResolutionException(
+                          "Received artifact value missing, failing pipeline")
+                      break
+                    case ExpectedArtifact.ArtifactField.MissingPolicy.EXPRESSION:
+                      if (!field.getExpression()) {
```

Comment:
Shouldn't this be handled already by the expected artifact's validate method?

Reply:
Yes, should I do a `NotImplemented` or something similar here instead?

Reply:
Well I don't think you need to do any of the validation work here, instead you can assume validation passed and treat it like that

Reply:
Removed spurious check, this section just handles whether the `Artifact` matches or not.